src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/FileBasedPrograms/CanonicalMiscFilesProject.cs

```diff
@@ -0,0 +1,297 @@
+                return _loadedProject;
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Failed to initialize canonical miscellaneous files project");
+                return null;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Adds a document to the canonical project.
+    /// </summary>
+    public async Task<Document?> AddDocumentAsync(
+        string filePath,
+        SourceText sourceText,
+        CancellationToken cancellationToken)
+    {
+        using (await _gate.DisposableWaitAsync(cancellationToken))
+        {
+            if (_loadedProject == null)
+            {
+                return null;
+            }
+
+            var projectFactory = _workspaceFactory.MiscellaneousFilesWorkspaceProjectFactory;
+            var workspace = projectFactory.Workspace;
+            var project = workspace.CurrentSolution.GetProject(_loadedProject.ProjectSystemProject.Id);
+            if (project == null)
+            {
+                _logger.LogError("Canonical project not found in workspace");
+                return null;
+            }
+
+            // Add the source file to the project
+            _loadedProject.ProjectSystemProject.AddSourceFile(filePath);
+
+            // Get the updated project and document
+            var updatedProject = workspace.CurrentSolution.GetProject(_loadedProject.ProjectSystemProject.Id);
+            var document = updatedProject?.Documents.FirstOrDefault(d => d.FilePath == filePath);
+
+            return document;
+        }
+    }
+
+    /// <summary>
+    /// Removes a document from the canonical project.
+    /// </summary>
+    public async Task<bool> RemoveDocumentAsync(string filePath, CancellationToken cancellationToken)
+    {
+        using (await _gate.DisposableWaitAsync(cancellationToken))
+        {
+            if (_loadedProject == null)
+            {
+                return false;
+            }
+
+            var projectFactory = _workspaceFactory.MiscellaneousFilesWorkspaceProjectFactory;
+            var workspace = projectFactory.Workspace;
+            var project = workspace.CurrentSolution.GetProject(_loadedProject.ProjectSystemProject.Id);
+            if (project == null)
+            {
+                return false;
+            }
+
+            var document = project.Documents.FirstOrDefault(d => d.FilePath == filePath);
+            if (document == null)
+            {
+                return false;
+            }
+
+            // Remove the source file from the project
+            _loadedProject.ProjectSystemProject.RemoveSourceFile(filePath);
+
+            return true;
+        }
+    }
+
+    private string CreateCanonicalProjectContent()
+    {
+        var artifactsPath = GetCanonicalProjectDirectory();
+        var targetFramework = Environment.GetEnvironmentVariable("DOTNET_RUN_FILE_TFM") ?? "net$(BundledNETCoreAppTargetFrameworkVersion)";
+
+        var virtualProjectXml = $"""
```

Comment:
I believe that since the canonical project and all its on-disk contents are in a temp directory, there shouldn't be any need to do most/all the virtual project stuff that is done for file-based apps. i.e. the project file used for `dotnet new classlib` should work here.